src/Microsoft.Framework.PackageManager/SourceControl/GitSourceControlProvider.cs

```diff
@@ -0,0 +1,178 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+
+namespace Microsoft.Framework.PackageManager.SourceControl
+{
+    internal class GitSourceControlProvider: SourceControlProvider
+    {
+        private const string RepositoryApp = "git";
+        
+        private const string RepositoryUrlKey = "url";
+        private const string CommitHashKey = "commit";
+        private const string ProjectPathKey = "path";
+
+        private bool? _isInstalled;
+
+        public GitSourceControlProvider(Reports buildReports)
+            :base(buildReports)
+        {
+        }
+
+        public override bool IsInstalled
+        {
+            get
+            {
+                if (!_isInstalled.HasValue)
+                {
+                    _isInstalled = ProcessUtilities.ExecutableExists(RepositoryApp);
+                }
+
+                return _isInstalled.Value;
+            }
+        }
+
+        public override void AddMissingSnapshotInformation(string folderName, IDictionary<string, string> snapshotInformation)
+        {
+            if (!snapshotInformation.ContainsKey(RepositoryUrlKey) || string.IsNullOrEmpty(snapshotInformation[RepositoryUrlKey]))
+            {
+                throw new ArgumentNullException("The repository URL must be specified.");
+            }
+
+            if (!snapshotInformation.ContainsKey(CommitHashKey))
+            {
+                snapshotInformation[CommitHashKey] = GetHeadCommitId(folderName);
+            }
+
+            if (!snapshotInformation.ContainsKey(ProjectPathKey))
+            {
+                var repoRoot = GetRepoRoot(folderName);
+
+                // Get the path relative to the repo root
+                var pathRelativeToRepoRoot = Path.GetDirectoryName(folderName)
+                    .Substring(repoRoot.Length)
+                    .Replace('\\', '/')
+                    .TrimStart('/');
+                snapshotInformation[ProjectPathKey] = pathRelativeToRepoRoot;
+            }
+        }
+
+        private string GetHeadCommitId(string folderName)
+        {
+            string stdOut;
+            string stdErr;
+
+            if (ProcessUtilities.RunApp(RepositoryApp, "rev-parse HEAD", folderName, out stdOut, out stdErr) &&
+                string.IsNullOrEmpty(stdErr))
+            {
+                return stdOut;
```

Comment:
It worth consider what if __git__ changes their output format how do we patch our tools? And can customer customize the logic to get around it.